src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,15 +300,44 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app max-command-failure? max-command-size]
+  (fn [req]
+    (if (= :post (:request-method req))
+      (let [length-in-bytes (request/content-length req)]
+
+        (if length-in-bytes
+          (log/infof "Processing command with a content-length of %s bytes" length-in-bytes )
+          (log/warn (str "No content length found for POST. "
+                         "POST bodies that are too large could cause memory-related failures.")))
+
+        (if (and length-in-bytes
+                 max-command-failure?
+                 (> length-in-bytes max-command-size))
+          (do
+            (log/warnf "content-length of command is %sMB and is larger than the maximum allowed command size of %sMB"
```

Comment:
This ends up looking like:

```
2015-12-08 20:54:29,940 WARN  [qtp752077813-86 - /pdb/cmd/v1?checksum=95f7fbca7440767b134a1ddb748db27552e0152d] [p.p.middleware] content-length of command is 6.180831MB and is larger than the maximum allowed command size of 1.0MB
```

I'd say, space between the float and 'MB', and maybe consider rounding that float to 2-3 digits. It just looks crazy with six digits like this, might as well be a byte :-).